src/conn/conn_dhandle.c

```diff
@@ -313,6 +313,15 @@ __wt_conn_dhandle_close(WT_SESSION_IMPL *session, bool final, bool mark_dead)
     btree = is_btree ? dhandle->handle : NULL;
 
     if (is_btree) {
+        /* Don't close the btree if there are changes not yet globally visible. */
+        /*
+         * !! Check if we have exclusive access here, else a transaction can start on this btree
+         * after this check !!
+         */
+        if (!WT_DHANDLE_IS_CHECKPOINT(dhandle) && (session != conn->sweep_session) &&
```

Comment:
Why session != conn->sweep_session is necessary?

Reply:
__wt_conn_dhandle_close() needs the sweep_session check because without it 3 subtests of one python test all fail in the same way and always in the sweep_session.

cd build
env WT_BUILDDIR=`pwd`/build python3 ../test/suite/run.py -j $(nproc) -v 4 test_hs21.py

[pid:291005]: FAIL: test_hs21.test_hs21.test_hs(column-fix)
and [pid:291005]: FAIL: test_hs21.test_hs21.test_hs(column)
and [pid:291005]: FAIL: test_hs21.test_hs21.test_hs(string-row)
Exact same output.  The output from one is below.
----------------------------------------------------------------------
[pid:291005]: testtools.testresult.real._StringException: Traceback (most recent call last):
  File "/home/ubuntu/wg/wt-10576-2/wiredtiger/test/suite/wttest.py", line 233, in _callTestMethod
    method()
  File "/home/ubuntu/wg/wt-10576-2/wiredtiger/test/suite/test_hs21.py", line 198, in test_hs
    self.assertGreaterEqual(final_dhandle_sweep_closes, self.numfiles)
  File "/usr/lib/python3.10/unittest/case.py", line 1250, in assertGreaterEqual
    self.fail(self._formatMessage(msg, standardMsg))
  File "/usr/lib/python3.10/unittest/case.py", line 675, in fail
    raise self.failureException(msg)
AssertionError: 0 not greater than or equal to 10

Line 198 that failed is: self.assertGreaterEqual(final_dhandle_sweep_closes, self.numfiles)

Thus without the sweep_session change the sweep server does not close any dhandles!

Reply:
__wt_conn_dhandle_close() needs the sweep_session check because without it 3 subtests of one python test all fail in the same way and always in the sweep_session.

`cd build
env WT_BUILDDIR=$(pwd)/build python3 ../test/suite/run.py -j $(nproc) -v 4 test_hs21.py

[pid:291005]: FAIL: test_hs21.test_hs21.test_hs(column-fix)
and [pid:291005]: FAIL: test_hs21.test_hs21.test_hs(column)
and [pid:291005]: FAIL: test_hs21.test_hs21.test_hs(string-row)
Exact same output.  The output from one is below.
----------------------------------------------------------------------
[pid:291005]: testtools.testresult.real._StringException: Traceback (most recent call last):
  File "/home/ubuntu/wg/wt-10576-2/wiredtiger/test/suite/wttest.py", line 233, in _callTestMethod
    method()
  File "/home/ubuntu/wg/wt-10576-2/wiredtiger/test/suite/test_hs21.py", line 198, in test_hs
    self.assertGreaterEqual(final_dhandle_sweep_closes, self.numfiles)
  File "/usr/lib/python3.10/unittest/case.py", line 1250, in assertGreaterEqual
    self.fail(self._formatMessage(msg, standardMsg))
  File "/usr/lib/python3.10/unittest/case.py", line 675, in fail
    raise self.failureException(msg)
AssertionError: 0 not greater than or equal to 10

Line 198 that failed is: self.assertGreaterEqual(final_dhandle_sweep_closes, self.numfiles)`

Thus without the sweep_session change the sweep server does not close any dhandles!

Reply:
Interesting. I did not think of any internal transactions when deciding this scheme (eg by checkpoints and sweep server). What is likely happening is that the sweep session and the checkpoint session are doing internal transactions. We might be computing and storing the latest txn-id for the involved handles when ```__txn_next_op()``` gets called by the checkpoint or sweep server thread.

If my intuition of what is happening is correct, we can just move the checkpoint and sweep session check to ```__txn_next_op()```, where we can store the latest txn-id only if changes come from an application thread. I will make another comment in ```__txn_next_op()``` to reflect that idea.

Reply:
The approach I suggest might also potentially resolve the situations where there are handles with uncommitted data that for some reason sweep is trying to close.

Reply:
The sweep server does not close any dhandle with uncommitted data. It only closes dhandles that do not have any modifications.  See below from __ sweep_close_dhandle_locked().
`    /* Only sweep clean trees. */
      if (btree != NULL && btree->modified)
           return (0);`

Reply:
`I don't know how to find the transactions on a table in order to roll them back. ` 
Transactions only live in sessions. Only the thread using a session can commit or rollback its transaction. There is no way for a random application thread doing a drop can rollback the transaction out from under another thread/session. There is no easy way to know, for any table, what operations in all sessions' transactions affect it.

Reply:
May be we should fix the tests instead of adding this check?

Reply:
bq. This changes the definition but, I think it is strange that an operation with a force flag can fail to succeed.

I agree it is strange. But I also think it is wrong to drop something with uncommitted updates on it. This means the other thread may crash after we drop the table. Not sure why we have the force option and when it is useful. From my understanding, the drop operation should always be conducted when it is safe so there is no need for the force option.

Reply:
The force option has misleading semantics/behaviour at the moment. Though the intuitive conclusion when we read "force" is to ignore all errors and make drop successful. I think it is only supposed to suppress a "file not found" error for a drop that has already succeeded. Making drop possible without checking for a previous attempt's conclusion.

```
    Config('force', 'false', r'''
        return success if the object does not exist''',
        type='boolean'),
```

I do not think that `force` is supposed to suppress all errors. If the file is busy or there are uncommitted updates, we should return `EBUSY`.

Reply:
Also, as a separate ticket, I propose we should rename "force" to something else, because it does cause this confusion. As part of that ticket, we should test and audit the code to ensure the API definition matches the behavior.

Reply:
To make it easier to see what is left I am going to resolve some conversations.